src/main/java/org/jenkinsci/plugins/mesos/api/Session.java

```diff
@@ -0,0 +1,160 @@
+import scala.compat.java8.OptionConverters;
+import scala.concurrent.ExecutionContext;
+
+/**
+ * A representation of the connection to Mesos.
+ *
+ * <p>This class should only be used by {@link org.jenkinsci.plugins.mesos.MesosApi}.
+ */
+public class Session {
+
+  private static final Logger logger = LoggerFactory.getLogger(Session.class);
+
+  // Interface to USI.
+  @Nonnull private final SourceQueueWithComplete<SchedulerCommand> commands;
+
+  public static Session create(
+      FrameworkInfo frameworkInfo,
+      MesosClientSettings clientSettings,
+      Optional<CredentialsProvider> provider,
+      SchedulerSettings schedulerSettings,
+      PodRecordRepository repository,
+      Settings operationalSettings,
+      Consumer<StateEventOrSnapshot> eventHandler,
+      BiFunction<Done, Throwable, Done> terminationHandler,
+      ExecutionContext context,
+      ActorSystem system,
+      ActorMaterializer materializer) {
+    Flow<SchedulerCommand, StateEvent, NotUsed> schedulerFlow =
+        RestartFlow.withBackoff(
+            Duration.ofSeconds(3),
+            Duration.ofSeconds(30),
+            0.2,
+            20,
+            () -> {
+              return connectClient(
+                      frameworkInfo,
+                      operationalSettings,
+                      clientSettings,
+                      provider,
+                      system,
+                      materializer)
+                  .thenCompose(
+                      client -> {
+                        final SchedulerFactory schedulerFactory =
+                            SchedulerFactory.create(
+                                client,
+                                repository,
+                                schedulerSettings,
+                                Metrics.getInstance(frameworkInfo.getName()),
+                                context);
+                        return Scheduler.asFlow(schedulerFactory);
+                      })
+                  .get()
+                  .getFlow(); // TODO: do not block
+            });
+
+    Pair<SourceQueueWithComplete<SchedulerCommand>, CompletionStage<Done>> pair =
+        runScheduler(operationalSettings, schedulerFlow, eventHandler, materializer);
+
+    // Handle stream termination.
+    // pair.second().handle(terminationHandler);
+
+    return new Session(pair.first());
+  }
+
+  public Session(SourceQueueWithComplete<SchedulerCommand> commands) {
+    this.commands = commands;
+  }
+
+  /** Establish a connection to Mesos via the v1 client. */
+  private static CompletableFuture<MesosClient> connectClient(
+      Protos.FrameworkInfo frameworkInfo,
+      Settings operationalSettings,
+      MesosClientSettings clientSettings,
+      Optional<CredentialsProvider> authorization,
+      ActorSystem system,
+      ActorMaterializer materializer) {
+
+    // TODO: move times and max retries into operational settings.
+    return RestartSource.onFailuresWithBackoff(
```

Comment:
I thinks this should be move into USI. The client should always try to reconnect if the initial connection fails. Please note that this is different from restarting the scheduler flow.